src/AppInstallerCommonCore/UserSettings/UserSettings.cpp

```diff
@@ -0,0 +1,132 @@
+    using namespace Runtime;
+    using namespace Utility;
+
+    static constexpr std::string_view s_SettingFileName = "settings.json"sv;
+    static constexpr std::string_view s_SettingBackupFileName = "settings.json.backup"sv;
+
+    static constexpr std::string_view s_SettingEmpty =
+        R"""({
+    // For documentation on these settings, see: https://aka.ms/winget-settings
+    // "source": {
+    //    "autoUpdateIntervalInMinutes": 5
+    // }
+})"""sv;
+
+    UserSettings::UserSettings() : m_type(UserSettingsType::Default)
+    {
+        Reload();
+    }
+
+    void UserSettings::Reload()
+    {
+        Json::Value settingsRoot = Json::Value::nullSingleton();
+        m_type = UserSettingsType::Default;
+
+        // Settings can be loaded from settings.json or settings.json.backup files.
+        // 1 - Use settings.json if exists and passes parsing.
+        // 2 - Use settings.backup.json if settings.json fails to parse.
+        // 3 - Use default (empty) if both settings files fail to load.
+
+        auto settingsJson = ParseFile(s_SettingFileName);
+        if (settingsJson.has_value())
+        {
+            m_type = UserSettingsType::Standard;
+            settingsRoot = settingsJson.value();
+        }
+
+        // Settings didn't parse or doesn't exist, try with backup.
+        if (settingsRoot.isNull())
+        {
+            auto settingsBackupJson = ParseFile(s_SettingBackupFileName);
+            if (settingsBackupJson.has_value())
+            {
+                m_warnings.emplace_back(SettingsWarnings::LoadedBackupSettings);
+                m_type = UserSettingsType::Backup;
+                settingsRoot = settingsBackupJson.value();
+            }
+        }
+
+        // Populate the settings.
+        auto sourceStr = Utility::ToString(Source::GetPropertyName());
+        auto source = std::make_unique<Source>(settingsRoot[sourceStr]);
+        auto sourceWarnings = source->Warnings();
+        std::move(sourceWarnings.begin(), sourceWarnings.end(), std::inserter(m_warnings, m_warnings.end()));
+
+        auto visualStr = Utility::ToString(Visual::GetPropertyName());
+        auto visual = std::make_unique<Visual>(settingsRoot[visualStr]);
+        auto visualWarnings = visual->Warnings();
+        std::move(visualWarnings.begin(), visualWarnings.end(), std::inserter(m_warnings, m_warnings.end()));
+
+        m_source = std::move(source);
+        m_visual = std::move(visual);
+    }
+
+    std::optional<Json::Value> UserSettings::ParseFile(const std::string_view& fileName)
+    {
+        auto stream = GetSettingStream(Type::UserFile, fileName);
+        if (stream)
+        {
+            Json::Value root;
+            Json::CharReaderBuilder builder;
+            const std::unique_ptr<Json::CharReader> reader(builder.newCharReader());
+
+            std::stringstream settingsContent;
+            settingsContent << stream->rdbuf();
+            std::string settingsContentStr = settingsContent.str();
```

Comment:
There is a utility function to read the entire stream already (ReadEntireStream).  I don't know if its implementation is better than this, but you should just use that function and replace the body if warranted.